[FLINK-4720] Implement archived version of the ExecutionGraph

zentol opened the pull request:
This PR allows an ExecutionGraph to be archived. The archived version is a serializable snapshot of the EG at the time it was archived.

To this end I implemented an archived version for the following classes:
* ~~ExecutionConfig~~
* ExecutionGraph
* ExecutionJobVertex
* ExecutionJobVertex
* Execution

The archived versions are simple containers. They retain all fields that were required by the WebInterface handlers and have appropriate getX() methods. There are no methods that change state, however the containers are not guaranteed to be immutable. They are created by calling archive() on the respective original instance.

To ensure compatibility with the existing handlers a common 'Access' interface was added for each of the above classes. This means that there is now an `AccessExecutionGraph` interface, which both `ExecutionGraph` and `ArchivedExecutionGraph` implement. The name is more or less a placeholder. Several handlers, relevant `JobManagerMessages` and tests were adjusted accordingly. Apart from the `JobVertexBackPressureHandler` all handlers now work on the `Access*` interfaces.

Furthermore, a number of methods were be added to the existing `Execution*` classes. These serve 2 purposes:
 1. avoid user-defined classes, since the WebInterface/History server will (soon&#x2122;) no longer have access to user classes
 2. prevent accesses upwards (i.e from Execution to ExecutionVertex), otherwise the archived structure would be insane :)

List of added methods:
* ~~ExecutionConfig:~~
 * ~~`getRestartStrategyAsString()`; added due to the possibility of user-defined restart strategies in the future as per FLINK-4596~~
 * ~~`getGlobalJobParametersMap()`; added since the `GlobalJobParameters` instances may be a user-defined class~~
* `ExecutionGraph`:
 * `getFailureCauseAsString()`; added since the `Throwable` may by a user-defined exception
 * ~~`getExecutionConfig()`, returns the deserialized `ExecutionConfig`; added since the user classloader is no longer available.~~
* `ExecutionJobVertex`:
 * `getName()`, returns `getJobVertex().getName()`; Added since `getJobVertex()` is no longer available.
 * `getCheckpointStats()`; Added since `getGraph()` is no longer available.
* `ExecutionVertex`:
 * `getFailureCauseAsString()`: see above
 * `getPriorExecutions()` (package private); required for construction of archived version
* `Execution`:
 * `getFailureCauseAsString()`; see above
 * `getParallelSubtaskIndex()`; added since `getVertex()` is no longer available.

The `prepareForArchiving` methods were removed since changes to the `ExecutionGraph` or others, for the purpose of archiving, are no longer necessary.

The `CheckpointStats` and `JobCheckpointStats` classes now implement the `Serializable` interface, so that we do not need a separate archived implementation for them as well.

zentol commented:
Rebasing right now...

mxm commented:
Thanks for the pr @zentol. Looks good to me! A couple of things I noticed: 

- The prefix of the interfaces (`Access*`) could be a bit confusing. What about these?
  - interface: `ExecutionGraph`
  - runtime:`RuntimeExecutionGraph`
  - archive: `ArchivedExecutionGraph` 

  That said, you probably chose to not rename the existing `ExecutionGraph` class to minimize changes.

- How about having an interface `Archiveable<T extends Serializable>` containing the `T archive()` method? This interface could be implemented by all runtime classes which are archievable.

- You chose not to port the `ExecutionConfig` or `ExecutionConfigSummary` to your changes while rebasing?

zentol commented:
I did not want to rename all usages of ExecutionGraph, correct. Otherwise i would've went with exactly the names you suggested.

I am using the ExecutionConfigSummary (and removed my original ArchivedExecutionConfig) when i rebased, so I'm not sure what you mean :/

mxm commented:
>I am using the ExecutionConfigSummary (and removed my original ArchivedExecutionConfig) when i rebased, so I'm not sure what you mean :/

That's what I meant, I wondered why you didn't use your archive design pattern for `ExecutionConfig`.

What do you think about the `Archiveable<T extends Serializable>` interface I proposed?

zentol commented:
Ah i see. I removed it since it was pretty much identical to the ExecutionConfigSummary class; it would've been a rename essentially. I like the interface idea, makes it a bit more organized.

zentol commented:
I've added the Archiveable interface. I also added it to the ExecutionConfig, which required moving the ExecutionConfigSummary to flink-core. Since this move destroys the history anyway i renamed it to ArchivedExecutionConfig for consistency purposes.

mxm commented:
Thanks. Looks good to merge!

zentol commented:
thank you for your review @mxm, will merge this now.
